### PR TITLE
If you can't shoot a gun, then... JUST TOSS IT AT THEM

### PR DESCRIPTION
**Pull request by paganiy**
## About The Pull Request
Adds a new trait TRAIT_TOSS_GUN_HARD (which currently comes from the Sleeping Carp scroll, ninja MODsuit, berserk armor, being a medieval pirate or highlander). The owner of this trait deals an additional **10-20** (depending on the size of the weapon) damage to the enemy by throwing a gun, and also knocks them for 0.5 seconds. If a ballistic weapon **with a magazine** was thrown, then the magazine is removed and the bolt is racked, and if **without a magazine**, then all cartridges (revolvers) or chambered one is removed. Here's a preview of how it works:

https://github.com/tgstation/tgstation/assets/126676387/2a56eb74-6d37-4343-a685-3bf7159ab628

## Why It's Good For The Game
It's classic
how-to-use-a-gun-in-katana-zero
## Changelog
:cl:
add: If you can't shoot a gun, then... JUST TOSS IT AT THEM (for 0.5 seconds of knockdown and 10-20 an additional brute damage)
/:cl:

### CODE REVIEW

**Comment by Singul0** [comment]
soul

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -1291,6 +1291,11 @@
 /atom/movable/hitby(atom/movable/hitting_atom, skipcatch, hitpush = TRUE, blocked, datum/thrownthing/throwingdatum)
 	if(HAS_TRAIT(src, TRAIT_NO_THROW_HITPUSH))
 		hitpush = FALSE
+	var/mob/living/carbon/thrower = throwingdatum?.get_thrower()
+	if(HAS_TRAIT(thrower, TRAIT_NOGUNS) && isgun(hitting_atom)) //throw a gun at them when they don't expect it
```
Comment (Jacquerel) [inline code/game/atoms_movable.dm:1295]: Is there no proc called by a thrown atom when it hits something? Ideally this should be there under the gun subtype, not performing a typecheck on everything that hits anything else.

**Review by Jacquerel — CHANGES_REQUESTED** [review]
This should probably be a separate trait from `TRAIT_NOGUN` and not always provided at the same time. Ideally traits just do the thing they say they do so people don't get surprised when they use what looks like a very simple trait and it turns out to have side effects.

**Comment by Iamgoofball** [comment]
If you wanted to make this extra cool, if it's a ballistic gun have them eject the magazine and clear the chamber too and toss that to the side like the HM2 Pacifist guy.

**Comment by Profakos** [comment]
Finally, people can become true Gunslingers.

**Comment by norsvenska** [comment]
"ammunition depleted? chuck the gun at your enemy" sillytip

File: code/game/atoms_movable.dm (hunk 1 above)
Comment (paganiy) [reply code/game/atoms_movable.dm]: I've done something. Is it fine?

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by TheVekter** [comment]
Give it to Wizards and let me throw the Mami spell guns at people.

**Comment by Jacquerel** [comment]
> Give it to Wizards and let me throw the Mami spell guns at people.

Unfortunately wands and staffs count as guns as far as the game is concerned

**Comment by TheVekter** [comment]
> > Give it to Wizards and let me throw the Mami spell guns at people.
> 
> Unfortunately wands and staffs count as guns as far as the game is concerned

You heard me.

File: code/modules/projectiles/gun.dm [hunk 2]
```diff
@@ -126,6 +126,30 @@
...
+
+	var/mob/living/carbon/thrower = throwingdatum?.get_thrower()
+	if(!HAS_TRAIT(thrower, TRAIT_TOSS_GUN_HARD))
+		return
+	if(!isliving(hit_atom))
+		return
+	var/mob/living/living_target = hit_atom
+	living_target.Knockdown(0.5 SECONDS)
+	living_target.apply_damage(src.w_class > WEIGHT_CLASS_NORMAL ? 15 : 10, BRUTE, thrower.zone_selected, wound_bonus = CANT_WOUND, attacking_item = src)
+	if(!istype(src, /obj/item/gun/ballistic))
```
[the remark this answers is not in the document]
Comment (paganiy) [reply code/modules/projectiles/gun.dm]: did it in a separate proc

File: code/modules/projectiles/gun.dm [hunk 3]
```diff
@@ -126,6 +126,30 @@
 	suppressed = null
 	update_appearance()
 
+/obj/item/gun/throw_impact(atom/hit_atom, datum/thrownthing/throwingdatum) //throw a gun at them. They don't expect it.
+	. = ..()
+
+	var/mob/living/carbon/thrower = throwingdatum?.get_thrower()
+	if(!HAS_TRAIT(thrower, TRAIT_TOSS_GUN_HARD))
```
[the remark this answers is not in the document]
Comment (paganiy) [reply code/modules/projectiles/gun.dm:138]: added a isnull check

**Pull request merged**